contracts/ibc-forwarder/src/contract.rs

```diff
@@ -87,9 +87,21 @@ pub fn execute(
     info: MessageInfo,
     msg: ExecuteMsg,
 ) -> NeutronResult<Response<NeutronMsg>> {
-    match msg {
-        ExecuteMsg::DistributeFallback { coins } => try_distribute_fallback(deps, env, info, coins),
-        ExecuteMsg::Tick {} => try_tick(deps, env, info),
+    match (msg, CONTRACT_STATE.load(deps.storage)?) {
+        // ticks coming in to ibc forwarder in Instantiaed state will attempt to register the ICA
+        (ExecuteMsg::Tick {}, ContractState::Instantiated) => {
+            verify_caller(&info.sender, &CONTRACT_OP_MODE.load(deps.storage)?)?;
+            try_register_ica(deps, env)
+        }
+        // ticks coming in to ibc forwarder in IcaCreated state will attempt to forward the funds
+        (ExecuteMsg::Tick {}, ContractState::IcaCreated) => {
+            verify_caller(&info.sender, &CONTRACT_OP_MODE.load(deps.storage)?)?;
+            try_forward_funds(env, deps)
+        }
+        // distributing the fallback split is permisionless and non state-dependent
+        (ExecuteMsg::DistributeFallback { coins }, _) => {
+            try_distribute_fallback(deps, env, info, coins)
+        }
```

Comment:
How about matching on the current state first, and then on the msg. Something like:
```rust
match (current_state()?, msg) {
  (ContractState::Instantiated, ExecuteMsg::Tick) => try_register_ica(deps, env),
  (ContractState::IcaCreated, ExecuteMsg::Tick) => try_forward_funds(deps, env),
  (ContractState::DistributeFallback { coins }, ExecuteMsg::Tick } => try_distribute_fallback(deps, env, info, coins),
  _ => Err(ContractOperationError::InvalidOperation),
}
```
Also, moving the `verify_caller` guards inside the respective state transition functions makes the `execute` function leaner.
Finally, any invalid state transition should return an error IMHO, wdyt ?

Reply:
Now, let's imagine that all contracts define an `allowed_state_transitions` map with the state as the key, and a list of allowed (execute) operations for that state, and that this map is exposed (can be queried). Then we don't need a generic Tick operation anymore.

Reply:
Very interesting suggestion @stiiifff. A benefit of exposing the state transition table through a query is that the contract becomes self-describing. We can then dynamically build a UI for this contract.

Reply:
agree on the matching the state first and moving the verifications into the functions, will add that. was mostly curious about the syntax for now.

regarding some formal definitions/boxes that would encapsulate the fsm logic, i'm inclined to hold that off for the time being. feels like it may be a little premature and potentially limiting in short/medium term, while i think those encapsulations will emerge naturally with time. wyt?